### ISSUE TITLE
Pure Data crashes when attempting to connect to Jack via Pipewire

### ISSUE THREAD

**Issue by mitchmindtree** [NixOS/nixpkgs#118835]
**Describe the bug**

Pure Data crashes with "buffer overflow detected" and signal 6 when selecting the "jack" audio backend while using pipewire.

---

**To Reproduce**

1. Enable pipewire like so:

   ```nix
   services.pipewire = {
     enable = true;
     alsa = {
       enable = true;
       support32Bit = true;
     };
     jack.enable = true;
     pulse.enable = true;
   };
   hardware.pulseaudio.enable = false;
   ```

2. Install `puredata` from nixpkgs as you wish.
3. Run `pd` to open Pure Data.
4. Select `Media > jack` to enable the jack audio engine.
5. Pure Data will immediately crash with:

   ```
   priority 8 scheduling failed.
   *** buffer overflow detected ***: terminated
   Pd: signal 6
   ```

---

**Expected behavior**

Pure Data doesn't crash when selecting jack.

In reality, my expectations are quite low considering how new pipewire is and how finicky pure data can be! Mostly I'm just posting this to see if anyone else has managed a workaround or if I can help to debug this further :smile: 

---

**Additional context**

```
$ pipewire --version
pipewire
Compiled with libpipewire 0.3.24
Linked with libpipewire 0.3.24
```

I just noticed that the pipewire package was just bumped to 0.3.25 today, but I haven't had a chance to try it just yet.

Currently packaged pure data version is 0.50.2 judging by the "About Pd" page in the GUI.

Both Ardour and Carla seem to be working nicely with Jack via Pipewire without any tweaking, so I'm thinking this must be some quirk specific to how Pure Data connects to Jack :thinking: 

Here is the output if I run `pd -d 5` (the highest debug level):

```
[mindtree@mindtree:~/programming/nix/mindtree-home]$ pd -d 5
pdtk_watchdog
set ::tmp_path {}
lappend ::tmp_path {/home/mindtree/pd/externals}
set ::sys_searchpath $::tmp_path
set ::tmp_path {}
set ::sys_temppath $::tmp_path
set ::tmp_path {}
lappend ::tmp_path {/home/mindtree/.local/lib/pd/extra/}
lappend ::tmp_path {/home/mindtree/pd-externals}
lappend ::tmp_path {/usr/local/lib/pd-externals}
lappend ::tmp_path {/home/mindtree/.nix-profile/lib/pd/extra}
set ::sys_staticpath $::tmp_path
set ::startup_flags {}
set ::startup_libraries {}
pdtk_pd_startup 0 50 2 {} { {ALSA 1} {jack 5} } {} {DejaVu Sans Mono} bold
set pd_whichapi 1
set zoom_open 0
::deken::set_platform Linux amd64 64 32
::pdwindow::post {priority 6 scheduling failed; running at normal priority
}
set pd_whichmidiapi 1
priority 8 scheduling failed.
*** buffer overflow detected ***: terminated
Pd: signal 6
```

Unfortunately most of the output is just the tracing of the startup process, and there's no new output when clicking "jack" which causes the crash (that part starts from "priority 8 scheduling failed").

If I monitor the "Messages" log in `qjackctl`, I can see there is a "JACK connection graph change" event when selecting "jack" in Pure Data, however I don't observe any change when watching the jack connection graph window.

When running and watching `pw-mon` the following occurs when selecting "jack" in Pure Data:

```
added:                                                                         
        id: 86                         
        permissions: rwxm                                                      
        type: PipeWire:Interface:Client (version 3)
        properties:                 
                pipewire.protocol = "protocol-native"
                pipewire.sec.pid = "20247"                
                pipewire.sec.uid = "1000"           
                pipewire.sec.gid = "100"        
                pipewire.sec.label = "kernel"        
                module.id = "2"                                                
                object.id = "86"                                               
                loop.cancel = "true"   
                client.name = "pure_data"            
                client.api = "jack"                                            
                config.name = "jack.conf"
                log.level = "0"
                context.spa-libs = "{support.* = "support/libspa-support"}"
                jack.properties = "{}"                                         
                clock.power-of-two-quantum = "true"
                default.clock.rate = "48000"
                default.clock.quantum = "1024"
                default.clock.min-quantum = "32"
                default.clock.max-quantum = "8192"
                default.video.width = "640"
                default.video.height = "480"
                default.video.rate.num = "25"
                default.video.rate.denom = "1"
                link.max-buffers = "64"
                mem.warn-mlock = "false"      
                mem.allow-mlock = "true"
                cpu.max-align = "64"                                           
                application.name = ".pd-wrapped"
                application.process.binary = ".pd-wrapped"
                application.language = "en_US.UTF-8"
                application.process.id = "20247"
                application.process.user = "mindtree"
                application.process.host = "mindtree"
                window.x11.display = ":0"
                core.version = "0.3.24"
                core.name = "pipewire-mindtree-20247"
                pipewire.access = "unrestricted"
added:
        id: 83
        permissions: rwxm
        type: PipeWire:Interface:Node (version 3)
        params:
        input ports: 0/1024
        output ports: 0/1024
        state: "suspended"
        properties:
                loop.cancel = "true"
                client.name = "pure_data"
                client.api = "jack"
                config.name = "jack.conf"
                node.name = "pure_data"
                node.description = "pure_data"
                media.type = "Audio"
                media.category = "Duplex"
                media.role = "DSP"
                node.always-process = "true"
                client.id = "86"
                object.id = "83"
removed:
        id: 83
removed:
        id: 86
``` 

I'm not yet familiar enough with pipewire to spot anything of concern in these add client and add node events, but perhaps someone else will spot something.

Related to #102547.

---

**Notify maintainers**

cc @cillianderoiste (`puredata` maintainer), @jansol (`pipewire` maintainer)

**Metadata**

```
$ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.18, NixOS, 21.05pre279456.04a2b269d89 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.05pre279456.04a2b269d89"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by jansol** [comment]
Works fine on 0.3.25 straight out of the box in `nix-shell -p puredata --run pd`. Tried enabling jack with 44.1kHz and 48kHz samplerates and confirmed that the sine wave and noise play correctly in the audio/midi test window.

**Comment by jansol** [comment]
Maybe related to this upstream issue that was fixed a short time ago? https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/988

**Comment by mitchmindtree** [comment]
Thanks for the tip @jansol!

I just updated my nixos unstable channel, switched to it and rebooted and can confirm I'm now running pipewire 0.3.25.

Unfortunately, I'm still getting the same crash as above when selecting jack. I've tried changing the sample rate, frames/period and the number of periods per buffer in qjackctl but this doesn't appear to help. I also tried changing the puredata audio settings to match the sample rate and frames/period shown in qjackctl but this does not appear to help either.

I'm curious what might be different between our setups. Any chance you could share the pipewire section of your configuration? Mine is still as above. Otherwise, I have a fairly stock GNOME desktop setup, and don't think I've got any other audio configuration.

Just to clarify, pure data opens and seems to run fine with ALSA (though uses a different audio device than what I have selected in my gnome audio settings) - it's only when switching to jack in the `media` menu that it crashes immediately.

**Comment by mitchmindtree** [comment]
In order to try and gain a little more context on what's going on, I've rebuilt the `puredata` package locally with the `--enable-debug` configure flag. Unfortunately this doesn't provide much more info when the crash occurs, so I thought I'd try `gdb`.

Using `gdb`, I attached to the running `.pd-wrapped` process. I set a breakpoint with `catch signal 6` in order to catch the process in the state where SIGABRT is raised. I then selected "Media > jack" in the GUI and gdb immediately caught it. Here's where the signal is caught, and the `backtrace` at that moment in execution:

```gdb
Thread 1 ".pd-wrapped" hit Catchpoint 1 (signal SIGABRT), 0x00007f3814dfa33a in raise () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
(gdb) backtrace
#0  0x00007f3814dfa33a in raise () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
#1  0x00007f3814de4523 in abort () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
#2  0x00007f3814e3a958 in __libc_message () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
#3  0x00007f3814ec82c2 in __fortify_fail () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
#4  0x00007f3814ec82a0 in __stack_chk_fail () from /nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6
#5  0x00000000004f0bb3 in jack_get_clients ()
#6  0x00000000004f125a in jack_open_audio ()
#7  0x000000000049e39c in sys_reopen_audio ()
#8  0x000000000049f487 in glob_audio_setapi ()
#9  0x0000000000498048 in pd_typedmess ()
#10 0x00000000004918a3 in binbuf_eval ()
#11 0x00000000004a28d3 in socketreceiver_read ()
#12 0x00000000004a1bb0 in sys_domicrosleep ()
#13 0x00000000004a349f in sys_pollgui ()
#14 0x000000000049d0a3 in m_pollingscheduler ()
#15 0x000000000049d268 in m_mainloop ()
#16 0x00000000004a6273 in sys_main ()
#17 0x00000000004170a2 in main ()
```

This seems to imply something is going wrong when pd attempts to build a list of jack clients in [this function](https://github.com/pure-data/pure-data/blob/master/src/s_audio_jack.c#L190-L256). The purpose of this function appears to populate the list of `jack_client_names` by iterating through the list of ports produced by `jack_get_ports`.

The first thing that jumps out to me in this function is that it seems to assume a jack client name cannot be longer than 100 `char`s long (see [here](https://github.com/pure-data/pure-data/blob/master/src/s_audio_jack.c#L206)). My Dell XPS 9310 appear to have some pretty huge client names, as shown in this graph:

![qjackctl-graph](https://user-images.githubusercontent.com/4587373/114312862-7b85dc00-9af4-11eb-87d0-be8c1ed46d20.png)

Indeed, the name of the bottom left audio node is:

```
Tiger Lake-LP Smart Sound Technology Audio Controller Headset Mono Microphone + Headphones Stereo Microphone
```

which is 108 bytes long.

I'll try recompile Pure Data with a larger name size limit to see if this helps, though I wouldn't be surprised if there are more places through the code base where magical/assumed size limits like this are used...

**Comment by mitchmindtree** [comment]
That seems to have fixed it!

More specifically, I changed the length of that temporary name char array from 100 to 128 and I'm no longer getting a crash when I select "Media -> jack". Pure data now shows up in the qjackctl graph and I can route audio etc as expected.

I'll close this in favour of opening a patch upstream, though with a better solution than just changing the magical length value.

I originally assumed this must be something to do with pipewire configuration, as I have run pure data on this device with JACK in the past before. However if I recall correctly, when I used to run JACK standalone (not via pipewire) it did not pick up as many clients as it does now with pipewire, so I never ran into this client name length issue in the past. Thanks anyways @jansol!
